File: mhc/mhc_pre_sinkhorn_backward/op_kernel/mhc_pre_sinkhorn_backward.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file mhc_pre_sinkhorn_backward.cpp
 * \brief
 */
#include "kernel_operator.h"
#include "lib/matmul_intf.h"
#include "arch22/mhc_pre_grad_kernel.h"
#include "arch22/mhc_pre_sinkhorn_backward_data_arch22.h"
#include "arch22/mhc_pre_sinkhorn_backward_key_arch22.h"

using namespace AscendC;

template<bool TILINGKEY>
__global__ __aicore__ void
mhc_pre_sinkhorn_backward(GM_ADDR grad_hin, GM_ADDR grad_h_post, GM_ADDR grad_h_res, GM_ADDR x, GM_ADDR phi,
                          GM_ADDR alpha, GM_ADDR bias, GM_ADDR h_pre, GM_ADDR hc_before_norm, GM_ADDR inv_rms,
                          GM_ADDR sum_out, GM_ADDR norm_out, GM_ADDR grad_x, GM_ADDR grad_phi, GM_ADDR grad_alpha,
                          GM_ADDR grad_bias, GM_ADDR workspace, GM_ADDR tiling)
{
    REGISTER_TILING_DEFAULT(MhcPreSinkhornBackwardArch22TilingData);
    GET_TILING_DATA_WITH_STRUCT(MhcPreSinkhornBackwardArch22TilingData, tilingData, tiling);
    GM_ADDR usrWorkspace = GetUserWorkspace(workspace);
    if (usrWorkspace == nullptr) {
        return;
    }
    KERNEL_TASK_TYPE_DEFAULT(KERNEL_TYPE_MIX_AIC_1_2);

    TPipe pipe;
    if constexpr (TILINGKEY) {
        MhcPreGradKernel<DTYPE_X, DTYPE_PHI, true> op;
        op.mm1_.SetSubBlockIdx(0);
        op.mm1_.Init(&tilingData.mm1TilingData, &pipe);
        op.mm2_.SetSubBlockIdx(0);
        op.mm2_.Init(&tilingData.mm2TilingData, &pipe);

        op.Init(x, phi, h_pre, grad_hin, grad_h_post, grad_h_res, alpha, bias, hc_before_norm, inv_rms, sum_out,
                norm_out, grad_x, grad_phi, grad_alpha, grad_bias, usrWorkspace, &tilingData, &pipe);
        op.Process();
    } else {
        MhcPreGradKernel<DTYPE_X, DTYPE_PHI, false> op;
        op.mm1_.SetSubBlockIdx(0);
        op.mm1_.Init(&tilingData.mm1TilingData, &pipe);
        op.mm2_.SetSubBlockIdx(0);
        op.mm2_.Init(&tilingData.mm2TilingData, &pipe);

        op.Init(x, phi, h_pre, grad_hin, grad_h_post, grad_h_res, alpha, bias, hc_before_norm, inv_rms, sum_out,
                norm_out, grad_x, grad_phi, grad_alpha, grad_bias, usrWorkspace, &tilingData, &pipe);
        op.Process();
    }
}

File: torch_extension/cann_ops_transformer/ops/mhc_pre_sinkhorn_backward.py
# -----------------------------------------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
import torch
import torch_npu
from torch.library import impl
from cann_ops_transformer.op_builder.builder import OpBuilder
from cann_ops_transformer.op_builder.builder import AS_LIBRARY


class MhcPreSinkhornBackwardOpBuilder(OpBuilder):
    def __init__(self):
        super(MhcPreSinkhornBackwardOpBuilder, self).__init__("mhc_pre_sinkhorn_backward")

    def sources(self):
        return ['ops/csrc/mhc_pre_sinkhorn_backward.cpp']

    def schema(self) -> str:
        return "mhc_pre_sinkhorn_backward(Tensor gradHin, Tensor gradHPost, Tensor gradHRes, " \
               "Tensor x, Tensor phi, Tensor alpha, Tensor bias, " \
               "Tensor hPre, Tensor hcBeforeNorm, Tensor invRms, " \
               "Tensor sumOut, Tensor normOut, float hcEps) -> " \
               "(Tensor, Tensor, Tensor, Tensor)"

    def register_meta(self):
        @impl(AS_LIBRARY, self.name, "Meta")
        def mhc_pre_sinkhorn_backward_meta(grad_hin, grad_h_post, grad_h_res,
                                                x, phi, alpha, bias,
                                                h_pre, hc_before_norm, inv_rms,
                                                sum_out, norm_out, hc_eps):
            grad_x = torch.empty_like(x)
            grad_phi = torch.empty_like(phi)
            grad_alpha = torch.empty_like(alpha)
            grad_bias = torch.empty_like(bias)
            return (grad_x, grad_phi, grad_alpha, grad_bias)


mhc_pre_sinkhorn_backward_op_builder = MhcPreSinkhornBackwardOpBuilder()
op_module = mhc_pre_sinkhorn_backward_op_builder.load()


@impl(AS_LIBRARY, mhc_pre_sinkhorn_backward_op_builder.name, "PrivateUse1")
def mhc_pre_sinkhorn_backward(grad_hin, grad_h_post, grad_h_res,
                                   x, phi, alpha, bias,
                                   h_pre, hc_before_norm, inv_rms,
                                   sum_out, norm_out, hc_eps):
    return op_module.mhc_pre_sinkhorn_backward(grad_hin, grad_h_post, grad_h_res,
                                                     x, phi, alpha, bias,
                                                     h_pre, hc_before_norm, inv_rms,
                                                     sum_out, norm_out, hc_eps)